Adding chart in Document of MS-Word File without reading Temp MS-Word File

sandeeptiwari32 opened the pull request:
Bug 61972 - Adding chart in Document of MS-Word File without reading Temp MS-Word File
this patch contains changes for creating XWPFChart object in XWPFDocument object without reading Temp ms-Word File which contains chart and then modifying them.
Changes Done as suggested

sandeeptiwari32 committed:
Adding support to create chart in ms-word

sandeeptiwari32 committed:
adding support of creating chart in ms-word

sandeeptiwari32 committed:
updated Required changes

updated Required changes

sandeeptiwari32 committed:
updated changes as suggested 

updated changes as suggested for Bug 61972 - Adding chart in Document of MS-Word File without reading Temp MS-Word File
this patch contains changes for creating XWPFChart object in XWPFDocument object without reading Temp ms-Word File which contains chart and then modifying them.

sandeeptiwari32 committed:
updated changes as suggested

updated changes as suggested

sandeeptiwari32 committed:
update changes as suggested

update changes as suggested

asfgit commented:
Can one of the admins verify this patch?

Alain-Bearez commented on src/ooxml/java/org/apache/poi/xwpf/usermodel/XWPFChart.java:
```diff
@@ -105,19 +136,218 @@ public boolean equals(Object obj) {
+     * @param inline this object is used to adjust the margin and dimension of chart 
+     * @since POI 4.0.0
+     */
+    protected void setInLine(CTInline ctInline) {
```
All methods using `CT*` classes should be annotated with `@Internal` to clearly state their purpose is not for general public use.

sandeeptiwari32 replied:
updated suggested changes

Alain-Bearez commented on src/ooxml/java/org/apache/poi/xwpf/usermodel/XWPFChartData.java:
```diff
@@ -0,0 +1,82 @@
+ * Raw picture data, normally attached to a WordprocessingML Drawing.
+ * As a rule, embedded xlsx are stored in the /word/embedded/ part of a WordprocessingML package.
+ */
+public class XWPFChartData extends POIXMLDocumentPart {
```
Add the `@Beta` annotation to the class.

sandeeptiwari32 replied:
updated suggested changes

Alain-Bearez commented on src/ooxml/java/org/apache/poi/xwpf/usermodel/XWPFChartData.java:
```diff
@@ -0,0 +1,82 @@
+     * Create a new XWPFChartData node
+     * @since POI 4.0.0
+     */
+    protected XWPFChartData() {
```
From the class name, I would have expected this class to be somewhat related to the `XDDFChartData` class. 

It also seems rather empty from behavior and only offers raw access to its input stream. Not sure what was your intent with that...

sandeeptiwari32 replied:
This class we are using for embedded part of xwpfchart object. Till now we are reading chart embedded part from template file and then updating its value according to use.
with above mention class we can write XSSFWorkbook to embedded part of chart directly

Alain-Bearez replied:
Then you don't need this class, see my other comment on the `XWPFRelation CHART_SHEET`.

Alain-Bearez commented on src/ooxml/java/org/apache/poi/xwpf/usermodel/XWPFDocument.java:
```diff
@@ -1603,9 +1603,69 @@ public XWPFTableCell getTableCell(CTTc cell) {
+        xwpfChart.addEmbeddedWorkSheet(xlsx.getDocumentPart().getPackagePart().getOutputStream());
+        
+        //set in line object into xwpfchart object
+        xwpfChart.setInLine(inline);
```
Maybe, instead of this `@Internal` method, you could define a public `attachTo` method accepting the `XWPFRun` as parameter, along with the boundaries and chart I'd.

sandeeptiwari32 replied:
Will update in next checkin

Alain-Bearez commented on src/ooxml/java/org/apache/poi/xwpf/usermodel/XWPFDocument.java:
```diff
@@ -1603,9 +1603,69 @@ public XWPFTableCell getTableCell(CTTc cell) {
+        CTInline inline = xRun.addChart(width, height,chartId);
+        
+        //get package part of xwpfchart object
+        XWPFChart xwpfChart=rp.getDocumentPart();
```
Check for consistently placing spaces around the `=` or after the `,`. This improves the readability for maintenance.

sandeeptiwari32 replied:
updated suggested changes

Alain-Bearez commented on src/ooxml/java/org/apache/poi/xwpf/usermodel/XWPFDocument.java:
```diff
@@ -1603,9 +1603,69 @@ public XWPFTableCell getTableCell(CTTc cell) {
+        xwpfChart.setInLine(inline);
+        
+        //add relation id in xwpfchart object
+        xwpfChart.getCTChartSpace().addNewExternalData().setId(xlsx.getRelationship().getId());
```
Seems like this should be handled by the `XDDFChart` class...

sandeeptiwari32 replied:
i searched for this in XDDFChart class but there is no any method for this purpose.I think your suggested change is that to create a method in XDDFchart for this purpose, if i am right then i will update it in next check in.

Alain-Bearez replied:
I will take some time to check out your branch and verify my hypothesis.

Alain-Bearez replied:
Something similar, but yet different, exists in `XSLFChart`. Sorry for my bad memories...

Alain-Bearez commented on src/ooxml/java/org/apache/poi/xwpf/usermodel/XWPFRun.java:
```diff
@@ -1024,7 +1026,64 @@ public XWPFPicture addPicture(InputStream pictureData, int pictureType, String f
+     * @throws IOException
+     * @since POI 4.0.0
+     */
+    public CTInline addChart(int width, int height,String chartRelId)
```
@Internal

sandeeptiwari32 replied:
updated suggested changes

Alain-Bearez commented on src/ooxml/testcases/org/apache/poi/xwpf/usermodel/TestXWPFChart.java:
```diff
@@ -87,4 +87,29 @@ public void testChartRelation() throws IOException
+        assertNotNull(chart.getCTChartSpace());
+        assertNotNull(chart.getCTChart());
+        assertEquals(XWPFChart.HEIGHT,chart.getChartHeight());
+        assertEquals(XWPFChart.WIDTH,chart.getChartWidth());
```
Constants have been renamed.

sandeeptiwari32 replied:
updated suggested changes

sandeeptiwari32 commented:
Sure sir....
I will update that comments...
And I am also incorporating changes of xddfchart method in xwpfchart class

sandeeptiwari32 committed:
Updated suggested changes.

Updated suggested changes.

sandeeptiwari32 committed:
Updated suggested changes.

Updated suggested changes.

sandeeptiwari32 committed:
Updated suggested changes.

Updated suggested changes.

sandeeptiwari32 committed:
Updated suggested changes.

Updated suggested changes.

sandeeptiwari32 committed:
updated suggested changes

updated suggested changes

sandeeptiwari32 committed:
updated suggested changes

updated suggested changes

sandeeptiwari32 reviewed (commented):
updated suggested changes.
please review with updated code

Alain-Bearez commented on src/ooxml/java/org/apache/poi/xwpf/usermodel/XWPFRelation.java:
```diff
@@ -28,120 +29,128 @@ Licensed to the Apache Software Foundation (ASF) under one or more
+            "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
+            POIXMLDocument.PACK_OBJECT_REL_TYPE,
+            "/word/embeddings/Microsoft_Excel_Worksheet#.xlsx",
+            XWPFChartData.class
```
You would not need to define `XWPFChartData` if you were to use `XSSFWorkbook.class` here.

Alain-Bearez commented on src/ooxml/java/org/apache/poi/xwpf/usermodel/XWPFRun.java:
```diff
@@ -1024,7 +1026,65 @@ public XWPFPicture addPicture(InputStream pictureData, int pictureType, String f
+            
+            org.w3c.dom.Document doc = DocumentHelper.readDocument(is);
+            
+            inline.set(XmlToken.Factory.parse(doc.getDocumentElement(), DEFAULT_XML_OPTIONS));
```
You would not need to pass the `chartRelId` as parameter if you were to place this initialization in `XWPFDocument.addChart`.

For information only, since I don't see yet how to refactor this, the same effect is achieved differently in `XSSFGraphicFrame.appendChartElement`.

Alain-Bearez commented on src/ooxml/java/org/apache/poi/xwpf/usermodel/XWPFRun.java:
```diff
@@ -1024,7 +1026,65 @@ public XWPFPicture addPicture(InputStream pictureData, int pictureType, String f
+            
+            CTDrawing drawing = run.addNewDrawing();
+            
+            CTInline inline = drawing.addNewInline();
```
Since variable `drawing` is not reused later on, you can inline its value here.

Alain-Bearez commented on src/ooxml/java/org/apache/poi/xwpf/usermodel/XWPFChart.java:
```diff
@@ -105,19 +137,219 @@ public boolean equals(Object obj) {
+     * @throws IOException
+     * @since POI 4.0.0
+     */
+    public boolean writeEmbeddedWorkSheet(Workbook wb) throws IOException {
```
I used a different approach in `XSLFChart`, but then I do not create a new chart from scratch there. If we were to find a common way of handling this, we might define an abstract class that extends from `XDDFChart` and is extended by both `XWPFChart` and `XSLFChart` but not `XSSFChart`.

sandeeptiwari32 replied:
yes sir we can do it for both XSLFChart and XWPFChart object in one common class.
but as for embedded part(.xlsx file) when we are using XSSFWorkbook.class in XSLF relation variable or in XWPFRelation.CHART_SHEET and calling saveWorkBook(XSSFWorkbook wb) method at that time it working fine. i mean we are able to write workbook to output stream.but when we are going to write final document object it was trying to write internal part of excel file eg. /doc/app.xml,/xl/workbook.xml etc. and due to that the generated document is corrupted, because of blank excel embedded part. While i am using XWPFChartData.class in relation and calling saveWorkbook method then it;s working fine. we need to look this issue.

Alain-Bearez commented on src/ooxml/java/org/apache/poi/xwpf/usermodel/XWPFDocument.java:
```diff
@@ -1603,9 +1603,69 @@ public XWPFTableCell getTableCell(CTTc cell) {
+     * @throws IOException
+     * @since POI 4.0.0
+     */
+    public XWPFChart addChart() throws InvalidFormatException, IOException{
```
Such methods are called `createChart` or `createPicture` or `createShape` in other place. I would prefer you to rename both your `addChart` methods to `createChart`.

Alain-Bearez commented on src/ooxml/java/org/apache/poi/xwpf/usermodel/XWPFDocument.java:
```diff
@@ -1603,9 +1603,69 @@ public XWPFTableCell getTableCell(CTTc cell) {
+        //create paragraph and run object
+        XWPFRun xRun = this.createParagraph().createRun();
+        
+        CTInline inline = xRun.addChart(width, height,chartId);
```
No need to pass `chartId` here, if you place the initialization (from `XWPFRun`, see comment there) in a private method `appendChartElement` in this class, where you will pass `inline` and `chartId` as arguments.

Alain-Bearez commented on src/ooxml/java/org/apache/poi/xwpf/usermodel/XWPFDocument.java:
```diff
@@ -1603,9 +1603,69 @@ public XWPFTableCell getTableCell(CTTc cell) {
+        RelationPart xlsx = xwpfChart.createRelationship(XWPFRelation.CHART_SHEET,chartNumber);
+        
+        //add output stream in xwpfchart object
+        xwpfChart.addEmbeddedWorkSheet(xlsx.getDocumentPart().getPackagePart().getOutputStream());
```
What about passing simply the workbook as argument?

```
xwpfChart.addEmbeddedWorkSheet(xlsx.getDocumentPart());
```

sandeeptiwari32 committed:
Updated suggested changes

Updated suggested changes

sandeeptiwari32 committed:
Updated suggested changes

Added common method of XWPFChart and XSLFChart classes

sandeeptiwari32 committed:
Updated suggested changes

removed common method from xlsfchart and update code for setting externaldata relation id, creating embedded and saving

sandeeptiwari32 committed:
Updated suggested changes

removed common method and moved to XDDFChart class

sandeeptiwari32 committed:
Updated suggested changes

changed addchart to createChart method

sandeeptiwari32 committed:
updated  poi version

updated  poi version

sandeeptiwari32 committed:
file not showing difference properly

file not showing difference properly

sandeeptiwari32 committed:
placed in wrong package

placed in wrong package

sandeeptiwari32 committed:
placed in wrong package

placed in wrong package

sandeeptiwari32 committed:
placed in wrong package

placed in wrong package

sandeeptiwari32 committed:
placed in wrong package

placed in wrong package

sandeeptiwari32 committed:
now not needed

now not needed

sandeeptiwari32 committed:
placed on required package

placed on required package

sandeeptiwari32 committed:
updated relation file for XSLF

updated relation file for XSLF

sandeeptiwari32 committed:
updated relation file for XWPF

updated relation file for XWPF

sandeeptiwari32 committed:
updated poi version

updated poi version

sandeeptiwari32 reviewed (commented):
updated suggested changes.
moved common method into XDDFChart class from XSLFChart and XWPFChart classes.
resolved overwriting of embedded data.

sandeeptiwari32 commented:
updated suggested changes.
moved common method into XDDFChart class from XSLFChart and XWPFChart classes.
resolved overwriting of embedded data.

Commit:
fix compile problems

sandeeptiwari32 committed:
merged import statement

merged import statement

sandeeptiwari32 committed:
merged import statement

merged import statement

sandeeptiwari32 committed:
Updated file to store workbook object.

Updated file to store workbook object while we are reading chart from template.

sandeeptiwari32 committed:
Updated file to store workbook object

Updated file to store workbook object while we are reading chart from template.

sandeeptiwari32 committed:
updated suggested changes.

updated suggested changes.

sandeeptiwari32 committed:
created constructor to hold embedded part.

created constructor to hold embedded part for xwpfchart and xslfchart while we are reading chart form template file.

sandeeptiwari32 committed:
change line end to LR

change line end to LR

sandeeptiwari32 reviewed (commented):
update code to handle XSSFWorkbook part while reading XWPFDocument,and also created method to holde embedded excel part so that at the time of writing document we can also save workbook part.
updated suggested chnages.

sandeeptiwari32 commented:
update code to handle XSSFWorkbook part while reading XWPFDocument,and also created method to holde embedded excel part so that at the time of writing document we can also save workbook part.
updated suggested chnages.

sandeeptiwari32 committed:
Merge pull request #1 from pjfanning/sandeep-trunk

@pjfanning compilation problem resolved. please check  with updated code

pjfanning commented:
I'm getting this test failure:

```<testcase classname="org.apache.poi.xwpf.usermodel.TestXWPFChart" name="testAddChartsToNewDocument" time="0.016">
    <error message="java.lang.reflect.InvocationTargetException" type="org.apache.poi.POIXMLException">org.apache.poi.POIXMLException: java.lang.reflect.InvocationTargetException
        at org.apache.poi.POIXMLFactory.createDocumentPart(POIXMLFactory.java:63)
        at org.apache.poi.POIXMLDocumentPart.read(POIXMLDocumentPart.java:670)
        at org.apache.poi.POIXMLDocumentPart.read(POIXMLDocumentPart.java:688)
        at org.apache.poi.POIXMLDocument.load(POIXMLDocument.java:164)
        at org.apache.poi.xwpf.usermodel.XWPFDocument.&lt;init&gt;(XWPFDocument.java:141)
        at org.apache.poi.xwpf.XWPFTestDataSamples.writeOutAndReadBack(XWPFTestDataSamples.java:42)
        at org.apache.poi.xwpf.usermodel.TestXWPFChart.testAddChartsToNewDocument(TestXWPFChart.java:114)
Caused by: java.lang.reflect.InvocationTargetException
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
        at org.apache.poi.xwpf.usermodel.XWPFFactory.createDocumentPart(XWPFFactory.java:57)
        at org.apache.poi.POIXMLFactory.createDocumentPart(POIXMLFactory.java:60)
Caused by: org.apache.poi.openxml4j.exceptions.NotOfficeXmlFileException: The supplied data appears to be a raw XML file. Formats such as Office 2003 XML are not supported
        at org.apache.poi.openxml4j.opc.internal.ZipHelper.verifyZipHeader(ZipHelper.java:174)
        at org.apache.poi.openxml4j.opc.internal.ZipHelper.openZipStream(ZipHelper.java:198)
        at org.apache.poi.openxml4j.opc.ZipPackage.&lt;init&gt;(ZipPackage.java:98)
        at org.apache.poi.openxml4j.opc.OPCPackage.open(OPCPackage.java:318)
        at org.apache.poi.util.PackageHelper.open(PackageHelper.java:36)
        at org.apache.poi.xssf.usermodel.XSSFWorkbook.&lt;init&gt;(XSSFWorkbook.java:296)
        at org.apache.poi.xssf.usermodel.XSSFWorkbook.&lt;init&gt;(XSSFWorkbook.java:338)
</error>```

pjfanning commented on src/ooxml/java/org/apache/poi/POIXMLDocumentPart.java:
```diff
@@ -50,12 +52,29 @@ Licensed to the Apache Software Foundation (ASF) under one or more
 public class POIXMLDocumentPart {
     private static final POILogger logger = POILogFactory.getLogger(POIXMLDocumentPart.class);
-
+    
```
could you change your IDE settings to not add these indents on blank lines?

pjfanning commented on src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFChart.java:
```diff
@@ -346,5 +374,224 @@ private void parseAxes() {
+        this.chart.set(other.chart);
+    }
+    
+    @Override
```
the annotation usually is added after the javadoc

sandeeptiwari32 committed:
Updated code for unit test case

Updated code for unit test case

sandeeptiwari32 committed:
Updated code for unit test case

Updated code for unit test case

sandeeptiwari32 committed:
Updated code for unit test case

Updated code for unit test case

sandeeptiwari32 committed:
changed file new line end

changed file new line end

sandeeptiwari32 committed:
changed file new line end

changed file new line end

sandeeptiwari32 committed:
changed file new line end

changed file new line end

sandeeptiwari32 commented:
@pjfanning updated suggested changes, as well as resolved unit test case failure.
i tried to changes IDE setting but still i am getting spaces in new line, it may be possible i am using windows environment.

sandeeptiwari32 reviewed (commented):
@pjfanning updated suggested changes, as well as resolved unit test case failure.
i tried to changes IDE setting but still i am getting spaces in new line, it may be possible i am using windows environment.

pjfanning commented:
LGTM. Merged.
